lib/mix/lib/mix/task.ex

```diff
@@ -380,12 +408,38 @@ defmodule Mix.Task do
 
     recursive = recursive(module)
 
+    {in_apps, apps} =
+      case options do
+        [apps: apps] -> {true, apps}
+        _ -> {false, nil}
+      end
+
     cond do
       recursive && Mix.Project.umbrella?() ->
         Mix.ProjectStack.recur(fn ->
-          recur(fn _ -> run(task, args) end)
+          if in_apps do
+            recur_in_apps(fn _ -> run(task, args) end, apps)
+          else
+            recur(fn _ -> run(task, args) end)
+          end
         end)
 
+      in_apps && !Mix.Project.umbrella?() ->
+        Mix.shell().info([
+          :yellow,
+          "warning: running #{inspect(task)} at root level because this is not an umbrella project"
+        ])
+
+        run(task, args)
+
+      in_apps && not recursive ->
```

Comment:
I think we should not care about this. Sometimes if the task is recursive is an implementation detail. We should just not emit any warning. :)